cannon/mipsevm/state_mt.go

```diff
@@ -0,0 +1,290 @@
+package mipsevm
+
+import (
+	"encoding/binary"
+	"fmt"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/common/hexutil"
+	"github.com/ethereum/go-ethereum/crypto"
+)
+
+// SerializedThreadSize is the size of a serialized ThreadContext object
+var SerializedThreadSize = 166
+
+// ThreadWitnessSize is the size of a thread witness encoded in bytes.
+//
+//	It consists of the active thread serialized and concatenated with the
+//	32 byte hash onion of the active thread stack without the active thread
+var ThreadWitnessSize = SerializedThreadSize + 32
+
+// The empty thread root - keccak256(bytes32(0) ++ bytes32(0))
+var EmptyThreadsRoot common.Hash = common.HexToHash("0xad3228b676f7d3cd4284a5443f17f1962b36e491b30a40b2405849e597ba5fb5")
+
+type ThreadContext struct {
+	// Metadata
+	ThreadId uint32 `json:"threadId"`
+	ExitCode uint8  `json:"exit"`
+	Exited   bool   `json:"exited"`
+	// State
+	FutexAddr        uint32     `json:"futexAddr"`
+	FutexVal         uint32     `json:"futexVal"`
+	FutexTimeoutStep uint64     `json:"futexTimeoutStep"`
+	Cpu              CpuScalars `json:"cpu"`
+	Registers        [32]uint32 `json:"registers"`
+}
+
+func (t *ThreadContext) serializeThread() []byte {
+	// TODO
+	out := make([]byte, SerializedThreadSize)
+
+	//out = binary.BigEndian.AppendUint32(out, t.Cpu.PC)
+	//out = binary.BigEndian.AppendUint32(out, t.Cpu.NextPC)
+	//out = binary.BigEndian.AppendUint32(out, t.Cpu.LO)
+	//out = binary.BigEndian.AppendUint32(out, t.Cpu.HI)
+	//
+	//for _, r := range t.Registers {
+	//	out = binary.BigEndian.AppendUint32(out, r)
+	//}
+
+	return out
+}
+
+func computeThreadRoot(prevStackRoot common.Hash, threadToPush *ThreadContext) common.Hash {
+	hashedThread := crypto.Keccak256Hash(threadToPush.serializeThread())
+
+	var hashData []byte
+	hashData = append(hashData, prevStackRoot[:]...)
+	hashData = append(hashData, hashedThread[:]...)
+
+	return crypto.Keccak256Hash(hashData)
+}
+
+// MT_STATE_WITNESS_SIZE is the size of the state witness encoding in bytes.
+const MT_STATE_WITNESS_SIZE = 155
+const (
+	MT_WITNESS_MEMROOT_OFFSET            = 0
+	MT_WITNESS_PREIMAGE_KEY_OFFSET       = MT_WITNESS_MEMROOT_OFFSET + 32
+	MT_WITNESS_PREIMAGE_OFFSET_OFFSET    = MT_WITNESS_PREIMAGE_KEY_OFFSET + 32
+	MT_WITNESS_HEAP_OFFSET               = MT_WITNESS_PREIMAGE_OFFSET_OFFSET + 4
+	MT_WITNESS_EXITCODE_OFFSET           = MT_WITNESS_HEAP_OFFSET + 4
+	MT_WITNESS_EXITED_OFFSET             = MT_WITNESS_EXITCODE_OFFSET + 1
+	MT_WITNESS_STEP_OFFSET               = MT_WITNESS_EXITED_OFFSET + 1
+	MT_WITNESS_WAKEUP_OFFSET             = MT_WITNESS_STEP_OFFSET + 8
+	MT_WITNESS_TRAVERSE_RIGHT_OFFSET     = MT_WITNESS_WAKEUP_OFFSET + 4
+	MT_WITNESS_LEFT_THREADS_ROOT_OFFSET  = MT_WITNESS_TRAVERSE_RIGHT_OFFSET + 1
+	MT_WITNESS_RIGHT_THREADS_ROOT_OFFSET = MT_WITNESS_LEFT_THREADS_ROOT_OFFSET + 32
+	MT_WITNESS_THREAD_ID_OFFSET          = MT_WITNESS_RIGHT_THREADS_ROOT_OFFSET + 32
+)
+
+type MTState struct {
```

Comment:
Open to other names for this 🤷‍♀️

Reply:
I think the answer here is to create some subpackages so we get better encapsulation, which will allow this to just be `State`.  I have another PR where I'm working on encapsulating some more targeted packages inside `mipsevm`.  So, I think this is just a temporary name.

Reply:
It's fine for now.

Reply:
Will the other PR be built on top of this one?

Reply:
yeah, I was going to rebase my work on top of this